### ISSUE TITLE
Syntax highlighting for nested square brackets

### ISSUE THREAD

**Issue by chrorn** [lervag/vimtex#2043]
See screenshot attached: the syntax highlighting when starting an environment with [] brackets that themselves contain (for instance) a citation with [] brackets seems like a bug to me. But I'm not sure---maybe I should just write this in a different way (however, latex compilation is without issues).
![Screenshot 2021-05-04 at 19 25 57](https://user-images.githubusercontent.com/24720490/117051934-00889d80-ad0f-11eb-9ff7-6a9cd66fe0ad.png)

**Comment by lervag** [comment]
Yes, I can see how you consider this a bug. Can you provide a full example of this that I can use to work on it?

**Comment by lervag** [comment]
Sorry: I did not mean "how you consider this a bug" as an offense, just to be clear. But before I fully agree I want to see a good example - I need to decide if this should be solved specifically for the package that defines `defn` or if it is more general.

**Comment by ji-cerny** [comment]
I see this as well. I believe it is pretty general when one uses theorem-like environments and needs to include a theorem from an older paper. Minimal example is
 
```tex
\documentclass{article}
\newtheorem{theorem}{Theorem}
\begin{document}
  \begin{theorem}[Foo {\cite[Theorem 1.1]{Bar20}}]
    Lorem ipsum \dots
  \end{theorem}
\end{document}
```
In the context of amsthm package, an example is given on page 4 of http://www.ams.org/arc/tex/amscls/amsthdoc.pdf, but it is not amsthm related issue. ``\begin{proof}`` has also such optional argument. 

Also, (unrealistic) code ``\cite[Lemma 1.1 {\cite[Foo]{Bar20}}]{Bar21}`` is highlighted correctly.

**Comment by chrorn** [comment]
Apologies for the late reply (also: absolutely no offense was taken). The example that ji-cerny mention is what I had in mind.

**Comment by lervag** [comment]
> In the context of amsthm package, an example is given on page 4 of http://www.ams.org/arc/tex/amscls/amsthdoc.pdf, but it is not amsthm related issue. `\begin{proof}` has also such optional argument.

Thanks for the example!

As you probably already understand, LaTeX syntax is inherently difficult to parse correctly. But, I think the solution in this case is to relax the default option groups to allow "TeX Groups", i.e. `{...}` that can contain top level LaTeX code. The current default option group includes standard commands, lengths, comments, but not the more "random" tex group. This change would mean that we suddenly "allow" `[...{tex stuff}...]` for most commands that take options (unless their already specialized). But I don't think this should cause any problems.

**Comment by clason** [comment]
(I think erring on the side of permissivity is good in general; I'd rather have something highlighted reasonably that is not correct TeX than have a correct (but not explicitly supported) special case incorrectly highlighted. The syntax highlighter should not moonlight as a parser, especially not for something as hard to parse as LaTeX ;))

**Comment by lervag** [comment]
> I think erring on the side of permissivity is good in general; I'd rather have something highlighted reasonably that is not correct TeX than have a correct (but not explicitly supported) special case incorrectly highlighted. The syntax highlighter should not moonlight as a parser, especially not for something as hard to parse as LaTeX ;)

@clason Thanks! I was actually considering to tag you just to get your opinion, but I realized I had already made up my mind (same conclusion). I also imagined you might already be aware of the discussion and pitch in if you had any thoughts.

I've pushed a change that I think solves this issue. Please let me know if this unexpectedly breaks something else!

**Comment by clason** [comment]
There's one regression from this: something like
```tex
    \begin{equation}
        \begin{cases}
            [0,1)
        \end{cases}
    \end{equation}
```
where the `[0,1)` is picked up as an option to the `cases` environment.

I think it'd make sense to tighten the regex so options are only recognized if they _immediately_ follow the environment? (So `\begin{cases}[foo]` but not `\begin{cases} [foo]`, and certainly not after a new line?)

**Comment by lervag** [comment]
Are you sure this is a regression? It seems to me that this is caused by changes made already in december last year.

Also, note that both whitespace and newlines are actually allowed in these cases. Still, I agree it seems a strange thing to do, and adding the restriction feels like an easy and OK way to solve this particular issue.

**Comment by clason** [comment]
You're right, it wasn't due to this PR but an earlier change. (I think we actually talked about this during the 2.0 PR iteration -- or maybe I just thought about bringing it up?)

I wasn't aware that this is legal; that changes things a bit. (Doing `\begin{cases}%` works as well, of course.)

Maybe as a compromise only make it more rigid in math mode (if that is possible)? I don't see many people typesetting intervals outside of math mode...

**Comment by lervag** [comment]
What I did right now was to make it more rigid only for environments. I can not really remember ever seeing anyone who adds space between `\begin{...}` and a following option group, so I believe this is quite safe. And if someone should complain, then I would not really mind arguing that it is much easier for them to just remove the space then it is for me to make a perfect syntax matcher. :\

We might have talked about this, yes. But I may have been to focussed on other things to actually make a full note of it.

But I agree that the compromise seems like a good approach. I'll implement it asap.

**Comment by lervag** [comment]
Ok, I made an attempt at it now, let me know if you find any further regressions.

**Comment by clason** [comment]
Looks good; will do! (As a proper issue instead of a drive-by comment ;))
